gno.land/pkg/gnoland/vals.go

```diff
@@ -0,0 +1,122 @@
+package gnoland
+
+import (
+	"fmt"
+	"strconv"
+
+	gnovm "github.com/gnolang/gno/gnovm/stdlibs/std"
+	abci "github.com/gnolang/gno/tm2/pkg/bft/abci/types"
+	"github.com/gnolang/gno/tm2/pkg/bft/types"
+	"github.com/gnolang/gno/tm2/pkg/crypto"
+	"github.com/gnolang/gno/tm2/pkg/events"
+)
+
+const (
+	valRealm = "gno.land/r/sys/vals"
+
+	validatorAddedEvent   = "ValidatorAdded"
+	validatorRemovedEvent = "ValidatorRemoved"
+
+	addressEventKey     = "address"
+	pubKeyEventKey      = "pub_key"
+	votingPowerEventKey = "voting_power"
+)
+
+// validatorEventFilter
+func validatorEventFilter(event events.Event) []abci.ValidatorUpdate {
```

Comment:
This is a potential evolution for issue GnoSDK V2 (post-MVP) #873. However, the initial version should not rely on `gnoEvent` for data transmission. Eventually, it can be used to trigger a hook, but then we need to rely on loading a `gno.Machine` and executing a function to receive the returned value. See GnoSDK MVP #2205.

Reply:
I see us going three routes:
- fully event based (this PR) model -> the most performant option IMO, it's completely async, relies on event parsing (events transfer data)
- hybrid model -> a wombo combo of events + running the VM. We still have the event switch logic, but now when the events are just "notifications" that the VM should be run, most probably to call `GetValidators()` from `r/sys/vals` (we would need a smart way to figure out removals / additions)
- fully VM-based model -> we run the VM on every `EndBlocker` to call `GetValidators()` from `r/sys/vals` even when there are no changes to the validator set (this is the slowest option of the 3)

Reply:
It's worth noting that we can always change / jump between these models. My strong preference is to keep it fully event-based, because as you can see in this PR it doesn't have a code toll we need to pay